### PR TITLE
Promotion from AMD internal branch for 2022.Q3.4

### PR DESCRIPTION
**Pull request by WenqingLiAMD**
@s-perron

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1cc4967ac525c97dffc6fb18b01189763d30865e
<details><summary>CTS tests (Failed: 0/291482)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.6, Gfx8</summary><ul>

<li>Passed: 37494/73926 (50.7%)</li>
<li>Failed: 0/73926 (0.0%)</li>
<li>Not Supported: 36432/73926 (49.3%)</li><li>Warnings: 0/73926 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73925 (52.1%)</li>
<li>Failed: 0/73925 (0.0%)</li>
<li>Not Supported: 35434/73925 (47.9%)</li><li>Warnings: 0/73925 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35388/73925 (47.9%)</li>
<li>Failed: 0/73925 (0.0%)</li>
<li>Not Supported: 38537/73925 (52.1%)</li><li>Warnings: 0/73925 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 36014/69706 (51.7%)</li>
<li>Failed: 0/69706 (0.0%)</li>
<li>Not Supported: 33692/69706 (48.3%)</li><li>Warnings: 0/69706 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/patch/PatchPreparePipelineAbi.cpp [hunk 1]
```diff
@@ -121,7 +121,6 @@ bool PatchPreparePipelineAbi::runImpl(Module &module, PipelineShadersResult &pip
   return true; // Modified the module.
 }
 
-// =====================================================================================================================
```
Comment (Flakebi) [inline lgc/patch/PatchPreparePipelineAbi.cpp:124]: Is this removed by accident?

**Review by Flakebi — COMMENTED** [review]
I didn’t look through all the raytracing code.
At least the typos need to be fixed, otherwise the CI will keep failing.
Fixing the formatting would be nice as well, but the CI should be able to cope with it if it’s not perfect.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 4591d9ee3041a2fde55a8ae7f045df80f80b95e0
<details><summary>CTS tests (Failed: 0/291482)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.6, Gfx8</summary><ul>

<li>Passed: 37494/73926 (50.7%)</li>
<li>Failed: 0/73926 (0.0%)</li>
<li>Not Supported: 36432/73926 (49.3%)</li><li>Warnings: 0/73926 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73925 (52.1%)</li>
<li>Failed: 0/73925 (0.0%)</li>
<li>Not Supported: 35434/73925 (47.9%)</li><li>Warnings: 0/73925 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35388/73925 (47.9%)</li>
<li>Failed: 0/73925 (0.0%)</li>
<li>Not Supported: 38537/73925 (52.1%)</li><li>Warnings: 0/73925 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 36014/69706 (51.7%)</li>
<li>Failed: 0/69706 (0.0%)</li>
<li>Not Supported: 33692/69706 (48.3%)</li><li>Warnings: 0/69706 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 5801624f878ee439d5ceafb73e37c73795349809
<details><summary>CTS tests (Failed: 0/291482)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.6, Gfx8</summary><ul>

<li>Passed: 37494/73926 (50.7%)</li>
<li>Failed: 0/73926 (0.0%)</li>
<li>Not Supported: 36432/73926 (49.3%)</li><li>Warnings: 0/73926 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73925 (52.1%)</li>
<li>Failed: 0/73925 (0.0%)</li>
<li>Not Supported: 35434/73925 (47.9%)</li><li>Warnings: 0/73925 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35388/73925 (47.9%)</li>
<li>Failed: 0/73925 (0.0%)</li>
<li>Not Supported: 38537/73925 (52.1%)</li><li>Warnings: 0/73925 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 36014/69706 (51.7%)</li>
<li>Failed: 0/69706 (0.0%)</li>
<li>Not Supported: 33692/69706 (48.3%)</li><li>Warnings: 0/69706 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — COMMENTED** [review]
What's your plan for all the `#if VKI_RAY_TRACING`? Most of them can (and should) be removed immediately.

**Review by s-perron — COMMENTED** [review]
The copyright information on the new files should be double checked.  It is odd that the years do not match for corresponding .cpp and .h files.  I'm guessing it was copy-paste and updated the final year.

I won't hold this up, but there is not a single new test written for a major new feature.  There is over 10,000 new lines of code, and there are no tests.  [Testing CTS is probably not good enough.](https://github.com/GPUOpen-Drivers/llpc/blob/dev/docs/Contributing.md#:~:text=We%20do%20not%20want%20to%20rely%20on%20the%20Vulkan%20CTS.%20Running%20the%20Vulkan%20CTS%20takes%20a%20long%20time%2C%20putting%20a%20time%2Dto%2Dtest%20burden%20on%20all%20developers.%20The%20Vulkan%20CTS%20tests%20are%20also%20harder%20to%20debug%20because%20more%20components%20are%20involved%20and%20require%20a%20GPU.)

I won't say anymore about the test.  If everybody at AMD agrees tests are not needed, then I will let it go.  If others at AMD feel tests are needed, then you can discuss among yourselves.

**Comment by amdrexu** [comment]
> I won't hold this up, but there is not a single new test written for a major new feature. There is over 10,000 new lines of code, and there are no tests. 

Hello Steven,
Let's explain this. The ray tracing feature is not  simple work that is done within several months. Rather, the efforts are 3+ years. During this period, the spec was changed and revised frequently and the spec became more and more complicated. Therefore, from the starting time of ray tracing, codes were changed all the time. It was very difficult to use unit tests to track the spec changes. And most ray tracing features were not independent. Instead, they interacted with each other. There are many CTS cases (50000+) regarding ray tracing, testing various scenes. Even if we use unit tests, the coverage is very very low compare to CTS and the tests are not concise enough to make a good matching pattern to verify the implementation of ray tracing. What complicates this is that the GPURT library is a shared one between teams (not only owned by LLPC team). Its frequent update easily causes the invalidation of existing unit tests even though the ray tracing spec is finalized. Of course, we value unit tests and believe it is right way in capturing small/isolated feature implementation and bug fixing. But for very complex features in Vulkan/OpenGL, we still have to resort CTS to verify each code submission. Hope this explains your concerns.

File: include/gpurt.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
+/*
+ ***********************************************************************************************************************
+ *
+ *  Copyright (c) 2019-2022 Advanced Micro Devices, Inc. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (jinjianrong) [reply include/gpurt.h:4]: The file was created internally in 2019.

File: include/vkgcDefs.h [hunk 3]
```diff
@@ -83,6 +83,9 @@
 //  | %Version | Change Description                                                                                    |
 //  | -------- | ------------------------------------------------------------------------------------------------------|
 //  |     54.6 | Add reverseThreadGroup to PipelineOptions                                                             |
+#if VKI_RAY_TRACING
+//  |     54.4 | Add isReplay to RayTracingPipelineBuildInfo for ray tracing capture replay feature                    |
+#endif
```
[the remark this answers is not in the document]
Comment (WenqingLiAMD) [reply include/vkgcDefs.h:88]: Hi Steven,

I believe the order is from internal code, version "54.5" is already added internally but not included in this release, which looks we are skipping some values.

Thanks,
Wenqing

File: lgc/patch/PatchPreparePipelineAbi.cpp (hunk 1 above)
Comment (WenqingLiAMD) [reply lgc/patch/PatchPreparePipelineAbi.cpp:124]: Yes, it is removed by accident, will change it back, man thanks~

File: lgc/state/PipelineState.cpp [hunk 4]
```diff
@@ -818,6 +818,7 @@ static bool nodeTypeHasBinding(ResourceNodeType nodeType) {
 
 // =====================================================================================================================
 // Check whether a (non-table) resource node matches the given {set,binding} compatible with nodeType
+//
```
[the remark this answers is not in the document]
Comment (WenqingLiAMD) [reply lgc/state/PipelineState.cpp:821]: Removed, many thanks~

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7a5bf5f2325c94447adf61ad30ab1d199a913ffb
<details><summary>CTS tests (Failed: 1/291482)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.6, Gfx8</summary><ul>

<li>Passed: 37494/73926 (50.7%)</li>
<li>Failed: 0/73926 (0.0%)</li>
<li>Not Supported: 36432/73926 (49.3%)</li><li>Warnings: 0/73926 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73925 (52.1%)</li>
<li>Failed: 0/73925 (0.0%)</li>
<li>Not Supported: 35434/73925 (47.9%)</li><li>Warnings: 0/73925 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35387/73925 (47.9%)</li>
<li><details><summary>Failed: 1/73925 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.texture.compressed.bc2_unorm_block_2d_npot_sparse
Stack trace: Flake
```
</p></details></li>

<li>Not Supported: 38537/73925 (52.1%)</li><li>Warnings: 0/73925 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 36014/69706 (51.7%)</li>
<li>Failed: 0/69706 (0.0%)</li>
<li>Not Supported: 33692/69706 (48.3%)</li><li>Warnings: 0/69706 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by jinjianrong** [comment]
> What's your plan for all the `#if VKI_RAY_TRACING`? Most of them can (and should) be removed immediately.

The purpose of keeping `#if VKI_RAY_TRACING` in the code is to make it easy for developers to check what code are for raytracing or  disable raytracing in the driver build (additional tools will be required to build raytracing, the build systems will need some time to adapt)

The macro will be removed in future release.

File: llpc/tool/llpcCompilationUtils.cpp [hunk 5]
```diff
@@ -397,6 +414,26 @@ Error processInputPipeline(ICompiler *compiler, CompileInfo &compileInfo, const
...
     }
   }
+#if VKI_RAY_TRACING
+  const BinaryData *shaderLibrary = nullptr;
+  if (pipelineState->pipelineType == VfxPipelineTypeRayTracing)
+    shaderLibrary = &pipelineState->rayPipelineInfo.shaderTraceRay;
+  else if (pipelineState->pipelineType == VfxPipelineTypeCompute)
+    shaderLibrary = &pipelineState->compPipelineInfo.shaderLibrary;
+#endif
+#if VKI_RAY_TRACING
```
[the remark this answers is not in the document]
Comment (WenqingLiAMD) [reply llpc/tool/llpcCompilationUtils.cpp:424]: Hi Steven, there exists some unreleased code between the two lines in our internal repo, so we can't remove it directly.

**Comment by amdvlk-admin** [comment]
### Test summary for commit a54249a17df49c95500108a732ebebc770710562
<details><summary>CTS tests (Failed: 0/291482)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.6, Gfx8</summary><ul>

<li>Passed: 37494/73926 (50.7%)</li>
<li>Failed: 0/73926 (0.0%)</li>
<li>Not Supported: 36432/73926 (49.3%)</li><li>Warnings: 0/73926 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73925 (52.1%)</li>
<li>Failed: 0/73925 (0.0%)</li>
<li>Not Supported: 35434/73925 (47.9%)</li><li>Warnings: 0/73925 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35388/73925 (47.9%)</li>
<li>Failed: 0/73925 (0.0%)</li>
<li>Not Supported: 38537/73925 (52.1%)</li><li>Warnings: 0/73925 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 36014/69706 (51.7%)</li>
<li>Failed: 0/69706 (0.0%)</li>
<li>Not Supported: 33692/69706 (48.3%)</li><li>Warnings: 0/69706 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
> I won't hold this up, but there is not a single new test written for a major new feature. There is over 10,000 new lines of code, and there are no tests. [Testing CTS is probably not good enough.](https://github.com/GPUOpen-Drivers/llpc/blob/dev/docs/Contributing.md#:~:text=We%20do%20not%20want%20to%20rely%20on%20the%20Vulkan%20CTS.%20Running%20the%20Vulkan%20CTS%20takes%20a%20long%20time%2C%20putting%20a%20time%2Dto%2Dtest%20burden%20on%20all%20developers.%20The%20Vulkan%20CTS%20tests%20are%20also%20harder%20to%20debug%20because%20more%20components%20are%20involved%20and%20require%20a%20GPU.)

FWIW, I'm torn on this question. It's clear that our testing has room for improvement.

I like leaning on CTS because it provides full end-to-end testing. However, our integration of CTS tests could be improved. For example, there could be a notification as soon as the first failure is detected, there could be automatic capturing of standard data (like TestResults.qpa and pipeline dumps), and so on. (Some of that is slightly better in our internal systems, but definitely not everything.)

Lit tests have value, too, but I feel that how we use them in LLPC/LGC also needs to be improved:

* Writing and updating test sources can be quite difficult because LGC-style IR is hard to read and write. I have some ideas for how to improve that (and some soon-to-be-published baby steps in that direction, though we really need core LLVM changes for the heavy hitters).
* It is not consistently practical to test individual passes, because many LGC-specific passes rely on sidecar data structures.
* Updating check lines in a meaningful way is difficult, and many of the checks are pretty low value in the first place. We'd ideally use some equivalent of `update_test_checks.py`

**Review by nhaehnle — APPROVED** [review]
As discussed there are some things that need to be cleaned up over time, but to make progress it seems best to land this first.

**Pull request closed without merging**